exact_string_matching/forward.py

```diff
@@ -29,3 +31,86 @@ def knuth_morris_pratt(t, w, n, m):
     if j == m:
       yield i
     i, j = i + j - sB[j], max(0, sB[j])
+
+@dataclass
+class _Hrp:
+  period: int
+  scope: Tuple[int, int]
+
+
+def _preprocess_first_hrps(w, m, k, limit = 2):
+  '''Compute list of first n k-Highly-Repeating-Prefixes (k-HRP).
+
+  A prefix is k-HRP if it is basic period with at least k periods.
+  '''
+  period, j = 1, 0
+  hrps: List[_Hrp] = []
+  while period + j < m:
+    while period + j < m and w[j + 1] == w[period + j + 1]:
+      j += 1
+    prefix_len = period + j
+
+    if period * k <= prefix_len:
+      hrps.append(_Hrp(period=period, scope=(2 * period, prefix_len)))
+      if len(hrps) == limit:
+        return hrps
```

Comment:
Consider changing it to yielding single values instead of returning a list with limit.